internal/controller/stages/control_flow_stages.go

```diff
@@ -0,0 +1,603 @@
+package stages
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/controller"
+	rollouts "github.com/akuity/kargo/internal/controller/rollouts/api/v1alpha1"
+	"github.com/akuity/kargo/internal/indexer"
+	"github.com/akuity/kargo/internal/kargo"
+	"github.com/akuity/kargo/internal/kubeclient"
+	libEvent "github.com/akuity/kargo/internal/kubernetes/event"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+type ControlFlowStageReconciler struct {
+	cfg           ReconcilerConfig
+	client        client.Client
+	eventRecorder record.EventRecorder
+}
+
+// NewControlFlowStageReconciler returns a new control flow Stage reconciler.
+// After creating the reconciler, call SetupWithManager to register it with a
+// controller manager.
+func NewControlFlowStageReconciler(
+	cfg ReconcilerConfig,
+) *ControlFlowStageReconciler {
+	return &ControlFlowStageReconciler{
+		cfg: cfg,
+	}
+}
+
+// SetupWithManager sets up the control flow Stage reconciler with the given
+// controller manager. It registers the reconciler with the manager and sets up
+// watches on the required objects.
+func (r *ControlFlowStageReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	// Configure client and event recorder using manager.
+	r.client = mgr.GetClient()
+	r.eventRecorder = libEvent.NewRecorder(ctx, mgr.GetScheme(), mgr.GetClient(), r.cfg.Name())
+
+	// This index is used to find all Freight that are directly available from
+	// a Warehouse. It is used to find Freight that can be sourced directly from
+	// the Warehouse for the control flow Stage.
+	if err := mgr.GetFieldIndexer().IndexField(
```

Comment:
This is a matter of personal preference, but I never liked the way the registration was factored out because it hid two important things from me when I wanted to know what a reconciler cared about:

1. The index field that's important to the reconciler.
2. The actual thing that decides how the value of this field is determined.

Given this, I took the liberty of not using the abstraction but rather doing the registration by hand here.

Reply:
Excellent rationale. 👍